{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "import pickle\n",
    "import phate\n",
    "import scprep\n",
    "import meld\n",
    "import time\n",
    "import graphtools as gt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import datetime\n",
    "import scanpy as sc\n",
    "from sklearn.decomposition import PCA\n",
    "import math\n",
    "from scipy import sparse\n",
    "import sys\n",
    "sys.path.append('/home/cl2292/project/')\n",
    "import xgboost as xgb\n",
    "import shap\n",
    "from sklearn import metrics\n",
    "from sklearn.model_selection import KFold\n",
    "from scipy.stats import zscore, binned_statistic\n",
    "from scipy.ndimage import gaussian_filter1d\n",
    "\n",
    "from hyperopt.pyll.stochastic import sample\n",
    "from timeit import default_timer as timer\n",
    "from hyperopt import tpe, Trials, fmin, hp, STATUS_OK\n",
    "\n",
    "\n",
    "# settings\n",
    "plt.rc('font', size = 9)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=9)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1\n",
    "sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak memory: 37628.87 MiB, increment: 37419.68 MiB\n",
      "loaded @220926.16:22:24\n",
      "took 35.53-s to load data\n",
      "peak memory: 73973.00 MiB, increment: 36351.47 MiB\n",
      "loaded @220926.16:22:59\n",
      "took 35.60-s to load data\n"
     ]
    }
   ],
   "source": [
    "# fps\n",
    "dfp = '/home/cl2292/project/SCA1_snRNAseq/Mouse/data/'\n",
    "pfp = '/home/cl2292/project/SCA1_snRNAseq/Mouse/results/'\n",
    "sc.settings.figdir = pfp\n",
    "\n",
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='220325_WT_imp.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit wtt = sc.read_h5ad(os.path.join(dfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))\n",
    "\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='220325_SCA1_imp.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit mutt = sc.read_h5ad(os.path.join(dfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def meld_pseudotime(AnnData, goi, one_df=False):\n",
    "    if AnnData.shape[0] > 5000:\n",
    "        AnnData = AnnData[AnnData.obs.sample(5000, replace=True).index, :]\n",
    "        AnnData.obs_names_make_unique() \n",
    "    AnnData = AnnData.copy() # to allow subsetting?\n",
    "    sc.pp.pca(AnnData)\n",
    "    sc.pp.neighbors(AnnData, n_pcs=50)\n",
    "\n",
    "    # MELD\n",
    "    G = gt.Graph(data=AnnData.obsp['connectivities']+sparse.diags([1]*AnnData.shape[0],format='csr'),\n",
    "                 precomputed='adjacency',\n",
    "                 use_pygsp=True)\n",
    "    G.knn_max = None\n",
    "    AnnData.obs['res_t']=-1\n",
    "    AnnData.obs.loc[AnnData.obs['timepoint']=='12wk', 'res_t'] = -0.5\n",
    "    AnnData.obs.loc[AnnData.obs['timepoint']=='18wk', 'res_t'] = 0\n",
    "    AnnData.obs.loc[AnnData.obs['timepoint']=='24wk', 'res_t'] = 0.5\n",
    "    AnnData.obs.loc[AnnData.obs['timepoint']=='30wk', 'res_t'] = 1\n",
    "    AnnData.obs['ees_t'] = meld.MELD().fit_transform(G=G, RES=AnnData.obs['res_t'])\n",
    "    AnnData.obs['ees_t'] = (AnnData.obs['ees_t'] - AnnData.obs['ees_t'].min()) / (AnnData.obs['ees_t'].max() - AnnData.obs['ees_t'].min())\n",
    "\n",
    "    # strata\n",
    "    X = pd.DataFrame(AnnData[:, goi].layers['imputed'], columns=goi, index=AnnData.obs.index.to_list())\n",
    "    y = AnnData.obs['ees_t'].to_numpy()\n",
    "    \n",
    "    if one_df:\n",
    "        # collate\n",
    "        X['genotype'] = AnnData.obs['genotype'].to_list()\n",
    "        X['timepoint'] = AnnData.obs['timepoint'].to_list()\n",
    "        X['Pseudotime'] = y\n",
    "        X['ctype'] = AnnData.obs['sub4'].to_list()\n",
    "        return X\n",
    "    else:\n",
    "        return X, y\n",
    "\n",
    "def dyn_gene_diff(WT, MUT, goi='all', gkey='sub4', gcname=None,\n",
    "                  pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/',\n",
    "                  normtype='ctype',\n",
    "                  ntimebins=20, verbose=True, out_file=None,\n",
    "                  mitoprefix='mt-', exclude_genes=['Xist', 'Tsix', 'Eif2s3y']):\n",
    "    '''\n",
    "    Arguments:\n",
    "      gkey (str) [optional, Default='ctype_ubcupdate']: .obs[key] specification\n",
    "        for grouping.\n",
    "      gcname (str or None) [optional, Default=None]: if None, loop through\n",
    "        all unique in c grouping. \n",
    "      normtype (str or None) [optional, Default=gtype_ctype]: standardize gex across genotype AND ctype groups. \n",
    "        Other options are None and ctype.\n",
    "    ''' \n",
    "    os.makedirs(pfp, exist_ok=True)\n",
    "    if goi == 'all':\n",
    "        goi = WT.var_names.to_list()\n",
    "        mito_genes = [i for i in WT.var_names if i.startswith(mitoprefix)]\n",
    "        exclude_genes = exclude_genes + mito_genes # this is kind of ignored with goi calls\n",
    "        goi = [i for i in goi if i not in exclude_genes]\n",
    "    results = pd.DataFrame()\n",
    "    \n",
    "    def ave_normedgex_pertbin_peradata(wt_gc, mut_gc, goi=goi, normtype=normtype, gcname = gcname, ntimebins=ntimebins):\n",
    "        # general, so automate\n",
    "        X_wt, y_wt = meld_pseudotime(wt_gc, goi)\n",
    "        X_mut, y_mut = meld_pseudotime(mut_gc, goi)\n",
    "        y = np.concatenate((y_wt, y_mut))\n",
    "\n",
    "        # binning\n",
    "        _, _, time_groupings = binned_statistic(y, y, bins=ntimebins)\n",
    "        \n",
    "        if normtype is not None:\n",
    "            print(normtype, 'here')\n",
    "            if normtype=='gtype_ctype':\n",
    "                X_wt = pd.DataFrame(zscore(X_wt), index=X_wt.index, columns=X_wt.columns)\n",
    "                X_mut = pd.DataFrame(zscore(X_mut), index=X_mut.index, columns=X_mut.columns)\n",
    "            elif normtype=='ctype':\n",
    "                X = X_wt.append(X_mut)\n",
    "                X = pd.DataFrame(zscore(X), index=X.index, columns=X.columns)\n",
    "                X_wt = X.loc[X_wt.index, :]\n",
    "                X_mut = X.loc[X_mut.index, :]\n",
    "            else:\n",
    "                raise NotImplementedError\n",
    "        X_wt['time_grouping'] = time_groupings[0:y_wt.shape[0]]\n",
    "        X_wt = X_wt.groupby('time_grouping').mean().T\n",
    "        X_mut['time_grouping'] = time_groupings[y_wt.shape[0]:]\n",
    "        X_mut = X_mut.groupby('time_grouping').mean().T\n",
    "        #DCN\n",
    "        if gcname == 'DCN':\n",
    "            X_mut[9] = (X_mut[10]+X_mut[8])/2\n",
    "            X_wt[9] = (X_wt[10]+X_wt[8])/2\n",
    "            \n",
    "        return X_wt, X_mut\n",
    "    \n",
    "    def make_plots(X1, X2, gcname, dX_df, out_file=out_file, ntimebins=ntimebins):\n",
    "        '''\n",
    "        Arguments:\n",
    "          X1: should be a dataframe with time_grouping in col index and values of elements\n",
    "            are average GEX values for gene indicated in index\n",
    "          X2: same as X1 but for mut AnnData obj\n",
    "        '''\n",
    "        cm = plt.get_cmap('magma')\n",
    "        timecolors = [cm(1.*i/ntimebins) for i in range(ntimebins)]\n",
    "        \n",
    "        # hmap1, WT\n",
    "        p = sns.clustermap(\n",
    "            X1,\n",
    "            pivot_kws=None,\n",
    "            method='average',\n",
    "            metric='euclidean',\n",
    "            z_score=None,\n",
    "            standard_scale=None,\n",
    "            row_cluster=True,\n",
    "            col_cluster=False,\n",
    "            row_linkage=None,\n",
    "            figsize=(4, 20),\n",
    "            col_linkage=None,\n",
    "            row_colors=None,\n",
    "            col_colors=timecolors,\n",
    "            mask=None,\n",
    "            dendrogram_ratio=0.2,\n",
    "            colors_ratio=0.03,\n",
    "            cbar_pos=(1, 0, 0.1, 0.15),\n",
    "            tree_kws=None,\n",
    "            cmap='RdYlBu_r',\n",
    "            yticklabels=True,\n",
    "            xticklabels=False,\n",
    "            vmax=1, vmin=-1,\n",
    "        )\n",
    "        p.ax_heatmap.set_xlabel('')\n",
    "        p.savefig(os.path.join(pfp, 'hmap_gexVtbin_wt_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "        \n",
    "        # hmap2, MUT\n",
    "        p = sns.clustermap(\n",
    "            X2,\n",
    "            pivot_kws=None,\n",
    "            method='average',\n",
    "            metric='euclidean',\n",
    "            z_score=None,\n",
    "            standard_scale=None,\n",
    "            row_cluster=True,\n",
    "            col_cluster=False,\n",
    "            row_linkage=None,\n",
    "            figsize=(4, 20),\n",
    "            col_linkage=None,\n",
    "            row_colors=None,\n",
    "            col_colors=timecolors,\n",
    "            mask=None,\n",
    "            dendrogram_ratio=0.2,\n",
    "            colors_ratio=0.03,\n",
    "            cbar_pos=(1, 0, 0.1, 0.15),\n",
    "            tree_kws=None,\n",
    "            cmap='RdYlBu_r',\n",
    "            yticklabels=True,\n",
    "            xticklabels=False,\n",
    "            vmax=1, vmin=-1,\n",
    "        )\n",
    "        p.ax_heatmap.set_xlabel('')\n",
    "        p.savefig(os.path.join(pfp, 'hmap_gexVtbin_mut_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "        \n",
    "        # hmap3, DIFF\n",
    "        dX  = X2.subtract(X1) # MUT - WT\n",
    "        p_diff = sns.clustermap(\n",
    "                dX,\n",
    "                pivot_kws=None,\n",
    "                method='average',\n",
    "                metric='euclidean',\n",
    "                z_score=None,\n",
    "                standard_scale=None,\n",
    "                row_cluster=True,\n",
    "                col_cluster=False,\n",
    "                row_linkage=None,\n",
    "                figsize=(4, 20), \n",
    "                col_linkage=None,\n",
    "                row_colors=None,\n",
    "                col_colors=timecolors,\n",
    "                mask=None,\n",
    "                dendrogram_ratio=0.2,\n",
    "                colors_ratio=0.03,\n",
    "                cbar_pos=(1, 0, 0.1, 0.15),\n",
    "                tree_kws=None,\n",
    "                cmap='RdYlBu_r',\n",
    "                yticklabels=True,\n",
    "                xticklabels=False,\n",
    "                vmax=1, vmin=-1,\n",
    "            )\n",
    "        p_diff.ax_heatmap.set_xlabel('')\n",
    "        p_diff.savefig(os.path.join(pfp, 'hmap_gexVtbin_DIFF_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "        \n",
    "        # regplot\n",
    "        x_diff_melted = dX.T.reset_index().melt(id_vars='time_grouping', var_name='GOI', value_name='mut-wt')\n",
    "        fig, ax = plt.subplots(1,1, figsize=(4,3))\n",
    "        p = sns.regplot('time_grouping', 'mut-wt', data=x_diff_melted, \n",
    "                    x_estimator=None, x_bins=None, x_ci='ci', scatter=True, \n",
    "                    fit_reg=True, ci=95, n_boot=1000, units=None, seed=None, \n",
    "                    order=1, logistic=False, lowess=True, robust=False, logx=False, x_partial=None, \n",
    "                    y_partial=None, truncate=True, dropna=True, x_jitter=0.5, y_jitter=None, \n",
    "                    label=None, color='gray', marker='o', scatter_kws={'s':1, 'alpha':0.6, 'lw':0}, line_kws=None, ax=ax)\n",
    "        fig.savefig(os.path.join(pfp, 'regplot_mut-wtVtbin_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "    \n",
    "        # histogram + smoothed line\n",
    "        fig, ax = plt.subplots(1,1, figsize=(4,3))\n",
    "        z = dX.abs().mean(axis=0).reset_index()\n",
    "        z = z.rename(columns={0:'ave_mut-wt'})\n",
    "        sns.barplot(x='time_grouping', y='ave_mut-wt', data=z, color='gray', ax=ax)\n",
    "        yprime = gaussian_filter1d(\n",
    "            z['ave_mut-wt'],\n",
    "            2,\n",
    "            axis=-1,\n",
    "            order=0,\n",
    "            output=None,\n",
    "            mode='nearest',\n",
    "            cval=0.0,\n",
    "            truncate=4.0,\n",
    "        )\n",
    "        ax.plot(z['time_grouping']-1, yprime, color='gray')\n",
    "        fig.savefig(os.path.join(pfp, 'hist_mut-wtABSDIFF_{}f'.format(gcname)), bbox_inches='tight')\n",
    "\n",
    "        dX = dX.reset_index()\n",
    "        dX['group'] = gcname\n",
    "        dX_df = dX_df.append(dX, ignore_index=True)\n",
    "        # intermediate save\n",
    "        if out_file is not None:\n",
    "            dX_df.to_csv(out_file)\n",
    "        return dX_df\n",
    "            \n",
    "    if gcname is None:\n",
    "        for ii, gc in enumerate(['GC','UBC','PC','MLI1','MLI2','GoC','AS','BG','OPC','OL','MG','PER','END','DCN']):\n",
    "            if gc=='GC': \n",
    "                # sample with replacement\n",
    "                nsample_idx = WT.obs.loc[(WT.obs[gkey]==gc), :].sample(5000).index.to_list()\n",
    "                wt = WT[(WT.obs[gkey]==gc) & (WT.obs.index.isin(nsample_idx))].copy()\n",
    "                nsample_idx = MUT.obs.loc[(MUT.obs[gkey]==gc), :].sample(5000).index.to_list()\n",
    "                mut = MUT[(MUT.obs[gkey]==gc) & (MUT.obs.index.isin(nsample_idx))].copy()\n",
    "            else:\n",
    "                wt = WT[(WT.obs[gkey]==gc), :].copy()\n",
    "                mut = MUT[(MUT.obs[gkey]==gc), :].copy()\n",
    "                \n",
    "            X_wt, X_mut = ave_normedgex_pertbin_peradata(wt, mut, gcname = gc)\n",
    "            results = make_plots(X_wt, X_mut, gc, results)\n",
    "            \n",
    "    else:\n",
    "        wt = WT[(WT.obs[gkey]==gcname), :].copy()\n",
    "        mut = MUT[(MUT.obs[gkey]==gcname), :].copy()\n",
    "        \n",
    "        X_wt, X_mut = ave_normedgex_pertbin_peradata(wt, mut, gcname = gcname)\n",
    "        results = make_plots(X_wt, X_mut, gcname, results)\n",
    "    \n",
    "    return results \n",
    "\n",
    "def diffplot_allgroups(x, ntimebins=20, sigma=2, gkey='group', pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/'):\n",
    "    cm = plt.get_cmap('magma')\n",
    "    timecolors = [cm(1.*i/ntimebins) for i in range(ntimebins)]\n",
    "    diffs = pd.DataFrame()\n",
    "    for i in ['GC','UBC','PC','MLI1','MLI2','GoC','AS','BG','OPC','OL','MG','PER','END','DCN']:\n",
    "        x_diff = x.loc[x[gkey]==i, [ii for ii in range(1, ntimebins+1)]+['index']].set_index('index')\n",
    "        \n",
    "\n",
    "        z = x_diff.abs().mean(axis=0).reset_index().rename(columns={'time_grouping':'time'})\n",
    "        z['time'] = z['time'].astype(int)\n",
    "        z = z.rename(columns={0:'ave_mut-wt'})\n",
    "\n",
    "        # smoothed\n",
    "        yprime = gaussian_filter1d(\n",
    "            z['ave_mut-wt'],\n",
    "            sigma,\n",
    "            axis=-1,\n",
    "            order=0,\n",
    "            output=None,\n",
    "            mode='nearest',\n",
    "            cval=0.0,\n",
    "            truncate=4.0,\n",
    "        )\n",
    "\n",
    "        dt = pd.DataFrame({'time':z['time'] - 1, 'ave_mut-wt':yprime})\n",
    "        dt[gkey] = i\n",
    "\n",
    "        diffs = diffs.append(dt, ignore_index=True)\n",
    "        \n",
    "    # aesthetics\n",
    "    cmap_ctype={'GC': '#FAC18A',\n",
    "            'DCN': '#AAABAB',\n",
    "            'UBC': '#BA61BA',\n",
    "            'PC': '#EE5264',\n",
    "            'MLI1': '#F9EBAE',\n",
    "            'MLI2': '#88BB92',\n",
    "            'GoC': '#46A928',\n",
    "            'AS': '#F9AEAE',\n",
    "            'BG': '#AEB7F9',\n",
    "            'OPC': '#F1815F',\n",
    "            'OL': '#75A3B7',\n",
    "            'MG': '#AC5861',\n",
    "            'PER': '#2D284B',\n",
    "            'END': '#1C67EE'}\n",
    "\n",
    "\n",
    "    fig, ax = plt.subplots(1,1, figsize=(5,4))\n",
    "    sns.lineplot(x='time', y='ave_mut-wt', hue=gkey, size=None, \n",
    "                     style=None, data=diffs, palette=cmap_ctype, hue_order=None, \n",
    "                     hue_norm=None, sizes=None, size_order=None, size_norm=None, \n",
    "                     dashes=True, markers=None, style_order=None, \n",
    "                     units=None, estimator='mean', ci=95, n_boot=1000, seed=None, sort=True, \n",
    "                     err_style='band', err_kws=None, legend='brief', ax=ax, lw=3)\n",
    "    ax.legend(bbox_to_anchor=(1.1,1)).set_title('')\n",
    "    ax.set_xticks([0,19])\n",
    "    fig.savefig(os.path.join(pfp, 'dyn_diffs_ctypes.pdf'), bbox_inches='tight')\n",
    "    return diffs\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# if __name__ == '__main__':\n",
    "\n",
    "#     wtt = sc.read('/home/cl2292/project/SCA1_snRNAseq/Mouse/data/220325_WT_imp.h5ad')\n",
    "#     wtt.obs = wtt.obs.merge(kwargs['add_md'], left_index=True, right_index=True)\n",
    "#     mutt = sc.read('/home/cl2292/project/SCA1_snRNAseq/Mouse/data/220325_SCA1_imp.h5ad')\n",
    "#     mutt.obs = mutt.obs.merge(kwargs['add_md'], left_index=True, right_index=True)\n",
    "#     scnddata.load_mouse_imputed_revision(add_md=scnddata.load_md())\n",
    "#     dX_df = dyn_gene_diff(wtt, mutt)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cic_binding = [\n",
    "    'Lrp1b',\n",
    "    'Nrg3',\n",
    "    'Dmd',\n",
    "    'Grid2',\n",
    "    'Ptprd',\n",
    "    'Dpyd',\n",
    "    'Cdk14',\n",
    "    'Lingo2',\n",
    "    'Cdh18',\n",
    "    'Sox5',\n",
    "    'Hgf',\n",
    "    'Robo2',\n",
    "    'Lrfn5',\n",
    "    'Ryr3',\n",
    "    'Kcnb2',\n",
    "    'Gpm6a',\n",
    "    'Dok6',\n",
    "    'Dgkb',\n",
    "    'Fgf12',\n",
    "    'Nrxn1',\n",
    "    'Grm1',\n",
    "    'Cntn5',\n",
    "    'Mdga2',\n",
    "    'Nebl',\n",
    "    'Cntn4',\n",
    "    'Dach1',\n",
    "    'Ctnna2',\n",
    "    'Megf10',\n",
    "    'Agbl4',\n",
    "    'Hmga2',\n",
    "    'Pcdh11x',\n",
    "    'Etv5',\n",
    "    'Lrp1b',\n",
    "    'Cntnap2',\n",
    "    'Gucy1a2',\n",
    "    'Lrrtm4',\n",
    "    'Atrnl1',\n",
    "    'Akt3',\n",
    "    'Macrod2',\n",
    "    'Khdrbs2',\n",
    "    'Cntnap5a',\n",
    "    'Utrn',\n",
    "    'Gpc5',\n",
    "    'Erbb4',\n",
    "    'Fgf14',\n",
    "    'Vps13b',\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dX_df = dyn_gene_diff(wtt, mutt, goi=cic_binding, normtype=None, pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/')\n",
    "diffs = diffplot_allgroups(dX_df, pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dyn_gene_diff_dcn(WT, MUT, goi='all', gkey='sub4', gcname=None,\n",
    "                  pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/',\n",
    "                  normtype='ctype',\n",
    "                  ntimebins=20, verbose=True, out_file=None,\n",
    "                  mitoprefix='mt-', exclude_genes=['Xist', 'Tsix', 'Eif2s3y']):\n",
    "    '''\n",
    "    Arguments:\n",
    "      gkey (str) [optional, Default='ctype_ubcupdate']: .obs[key] specification\n",
    "        for grouping.\n",
    "      gcname (str or None) [optional, Default=None]: if None, loop through\n",
    "        all unique in c grouping. \n",
    "      normtype (str or None) [optional, Default=gtype_ctype]: standardize gex across genotype AND ctype groups. \n",
    "        Other options are None and ctype.\n",
    "    ''' \n",
    "    os.makedirs(pfp, exist_ok=True)\n",
    "    if goi == 'all':\n",
    "        goi = WT.var_names.to_list()\n",
    "        mito_genes = [i for i in WT.var_names if i.startswith(mitoprefix)]\n",
    "        exclude_genes = exclude_genes + mito_genes # this is kind of ignored with goi calls\n",
    "        goi = [i for i in goi if i not in exclude_genes]\n",
    "    results = pd.DataFrame()\n",
    "    \n",
    "    def ave_normedgex_pertbin_peradata(wt_gc, mut_gc, goi=goi, normtype=normtype, ntimebins=ntimebins):\n",
    "        # general, so automate\n",
    "        X_wt, y_wt = meld_pseudotime(wt_gc, goi)\n",
    "        X_mut, y_mut = meld_pseudotime(mut_gc, goi)\n",
    "        y = np.concatenate((y_wt, y_mut))\n",
    "\n",
    "        # binning\n",
    "        _, _, time_groupings = binned_statistic(y, y, bins=ntimebins)\n",
    "        \n",
    "        if normtype is not None:\n",
    "            print(normtype, 'here')\n",
    "            if normtype=='gtype_ctype':\n",
    "                X_wt = pd.DataFrame(zscore(X_wt), index=X_wt.index, columns=X_wt.columns)\n",
    "                X_mut = pd.DataFrame(zscore(X_mut), index=X_mut.index, columns=X_mut.columns)\n",
    "            elif normtype=='ctype':\n",
    "                X = X_wt.append(X_mut)\n",
    "                X = pd.DataFrame(zscore(X), index=X.index, columns=X.columns)\n",
    "                X_wt = X.loc[X_wt.index, :]\n",
    "                X_mut = X.loc[X_mut.index, :]\n",
    "            else:\n",
    "                raise NotImplementedError\n",
    "        X_wt['time_grouping'] = time_groupings[0:y_wt.shape[0]]\n",
    "        X_wt = X_wt.groupby('time_grouping').mean().T\n",
    "        X_mut['time_grouping'] = time_groupings[y_wt.shape[0]:]\n",
    "        X_mut = X_mut.groupby('time_grouping').mean().T\n",
    "        return X_wt, X_mut\n",
    "    \n",
    "    def make_plots(X1, X2, gcname, dX_df, out_file=out_file, ntimebins=ntimebins):\n",
    "        '''\n",
    "        Arguments:\n",
    "          X1: should be a dataframe with time_grouping in col index and values of elements\n",
    "            are average GEX values for gene indicated in index\n",
    "          X2: same as X1 but for mut AnnData obj\n",
    "        '''\n",
    "        cm = plt.get_cmap('magma')\n",
    "        timecolors = [cm(1.*i/ntimebins) for i in range(ntimebins)]\n",
    "        \n",
    "        # hmap1, WT\n",
    "        p = sns.clustermap(\n",
    "            X1,\n",
    "            pivot_kws=None,\n",
    "            method='average',\n",
    "            metric='euclidean',\n",
    "            z_score=None,\n",
    "            standard_scale=None,\n",
    "            row_cluster=True,\n",
    "            col_cluster=False,\n",
    "            row_linkage=None,\n",
    "            figsize=(4, 20),\n",
    "            col_linkage=None,\n",
    "            row_colors=None,\n",
    "            col_colors=timecolors,\n",
    "            mask=None,\n",
    "            dendrogram_ratio=0.2,\n",
    "            colors_ratio=0.03,\n",
    "            cbar_pos=(1, 0, 0.1, 0.15),\n",
    "            tree_kws=None,\n",
    "            cmap='RdYlBu_r',\n",
    "            yticklabels=True,\n",
    "            xticklabels=False,\n",
    "            vmax=1, vmin=-1,\n",
    "        )\n",
    "        p.ax_heatmap.set_xlabel('')\n",
    "        p.savefig(os.path.join(pfp, 'hmap_gexVtbin_wt_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "        \n",
    "        # hmap2, MUT\n",
    "        p = sns.clustermap(\n",
    "            X2,\n",
    "            pivot_kws=None,\n",
    "            method='average',\n",
    "            metric='euclidean',\n",
    "            z_score=None,\n",
    "            standard_scale=None,\n",
    "            row_cluster=True,\n",
    "            col_cluster=False,\n",
    "            row_linkage=None,\n",
    "            figsize=(4, 20),\n",
    "            col_linkage=None,\n",
    "            row_colors=None,\n",
    "            col_colors=timecolors,\n",
    "            mask=None,\n",
    "            dendrogram_ratio=0.2,\n",
    "            colors_ratio=0.03,\n",
    "            cbar_pos=(1, 0, 0.1, 0.15),\n",
    "            tree_kws=None,\n",
    "            cmap='RdYlBu_r',\n",
    "            yticklabels=True,\n",
    "            xticklabels=False,\n",
    "            vmax=1, vmin=-1,\n",
    "        )\n",
    "        p.ax_heatmap.set_xlabel('')\n",
    "        p.savefig(os.path.join(pfp, 'hmap_gexVtbin_mut_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "        \n",
    "        # hmap3, DIFF\n",
    "        dX  = X2.subtract(X1) # MUT - WT\n",
    "        p_diff = sns.clustermap(\n",
    "                dX,\n",
    "                pivot_kws=None,\n",
    "                method='average',\n",
    "                metric='euclidean',\n",
    "                z_score=None,\n",
    "                standard_scale=None,\n",
    "                row_cluster=True,\n",
    "                col_cluster=False,\n",
    "                row_linkage=None,\n",
    "                figsize=(4, 20), \n",
    "                col_linkage=None,\n",
    "                row_colors=None,\n",
    "                col_colors=timecolors,\n",
    "                mask=None,\n",
    "                dendrogram_ratio=0.2,\n",
    "                colors_ratio=0.03,\n",
    "                cbar_pos=(1, 0, 0.1, 0.15),\n",
    "                tree_kws=None,\n",
    "                cmap='RdYlBu_r',\n",
    "                yticklabels=True,\n",
    "                xticklabels=False,\n",
    "                vmax=1, vmin=-1,\n",
    "            )\n",
    "        p_diff.ax_heatmap.set_xlabel('')\n",
    "        p_diff.savefig(os.path.join(pfp, 'hmap_gexVtbin_DIFF_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "        \n",
    "        # regplot\n",
    "        x_diff_melted = dX.T.reset_index().melt(id_vars='time_grouping', var_name='GOI', value_name='mut-wt')\n",
    "        fig, ax = plt.subplots(1,1, figsize=(4,3))\n",
    "        p = sns.regplot('time_grouping', 'mut-wt', data=x_diff_melted, \n",
    "                    x_estimator=None, x_bins=None, x_ci='ci', scatter=True, \n",
    "                    fit_reg=True, ci=95, n_boot=1000, units=None, seed=None, \n",
    "                    order=1, logistic=False, lowess=True, robust=False, logx=False, x_partial=None, \n",
    "                    y_partial=None, truncate=True, dropna=True, x_jitter=0.5, y_jitter=None, \n",
    "                    label=None, color='gray', marker='o', scatter_kws={'s':1, 'alpha':0.6, 'lw':0}, line_kws=None, ax=ax)\n",
    "        fig.savefig(os.path.join(pfp, 'regplot_mut-wtVtbin_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "    \n",
    "        # histogram + smoothed line\n",
    "        fig, ax = plt.subplots(1,1, figsize=(4,3))\n",
    "        z = dX.abs().mean(axis=0).reset_index()\n",
    "        z = z.rename(columns={0:'ave_mut-wt'})\n",
    "        sns.barplot(x='time_grouping', y='ave_mut-wt', data=z, color='gray', ax=ax)\n",
    "        yprime = gaussian_filter1d(\n",
    "            z['ave_mut-wt'],\n",
    "            2,\n",
    "            axis=-1,\n",
    "            order=0,\n",
    "            output=None,\n",
    "            mode='nearest',\n",
    "            cval=0.0,\n",
    "            truncate=4.0,\n",
    "        )\n",
    "        ax.plot(z['time_grouping']-1, yprime, color='gray')\n",
    "        fig.savefig(os.path.join(pfp, 'hist_mut-wtABSDIFF_{}'.format(gcname)), bbox_inches='tight')\n",
    "\n",
    "        dX = dX.reset_index()\n",
    "        dX['group'] = gcname\n",
    "        dX_df = dX_df.append(dX, ignore_index=True)\n",
    "        # intermediate save\n",
    "        if out_file is not None:\n",
    "            dX_df.to_csv(out_file)\n",
    "        return dX_df\n",
    "            \n",
    "    if gcname is None:\n",
    "        for ii, gc in enumerate(['GC','UBC','PC','MLI1','MLI2','GoC','AS','BG','OPC','OL','MG','PER','END','DCN']):\n",
    "            if gc=='GC': \n",
    "                # sample with replacement\n",
    "                nsample_idx = WT.obs.loc[(WT.obs[gkey]==gc), :].sample(5000).index.to_list()\n",
    "                wt = WT[(WT.obs[gkey]==gc) & (WT.obs.index.isin(nsample_idx))].copy()\n",
    "                nsample_idx = MUT.obs.loc[(MUT.obs[gkey]==gc), :].sample(5000).index.to_list()\n",
    "                mut = MUT[(MUT.obs[gkey]==gc) & (MUT.obs.index.isin(nsample_idx))].copy()\n",
    "            else:\n",
    "                wt = WT[(WT.obs[gkey]==gc), :].copy()\n",
    "                mut = MUT[(MUT.obs[gkey]==gc), :].copy()\n",
    "                \n",
    "            X_wt, X_mut = ave_normedgex_pertbin_peradata(wt, mut)\n",
    "            results = make_plots(X_wt, X_mut, gc, results)\n",
    "            \n",
    "    else:\n",
    "        wt = WT[(WT.obs[gkey]==gcname), :].copy()\n",
    "        mut = MUT[(MUT.obs[gkey]==gcname), :].copy()\n",
    "        \n",
    "        X_wt, X_mut = ave_normedgex_pertbin_peradata(wt, mut)\n",
    "        results = make_plots(X_wt, X_mut, gcname, results)\n",
    "    \n",
    "    return results \n",
    "\n",
    "def diffplot_dcn(x, ntimebins=20, sigma=2, gkey='group', pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/'):\n",
    "    cm = plt.get_cmap('magma')\n",
    "    timecolors = [cm(1.*i/ntimebins) for i in range(ntimebins)]\n",
    "    diffs = pd.DataFrame()\n",
    "    for i in ['DCN']:\n",
    "        x_diff = x.loc[x[gkey]==i, [ii for ii in range(1, ntimebins+1)]+['index']].set_index('index')\n",
    "        \n",
    "\n",
    "        z = x_diff.abs().mean(axis=0).reset_index().rename(columns={'time_grouping':'time'})\n",
    "        z['time'] = z['time'].astype(int)\n",
    "        z = z.rename(columns={0:'ave_mut-wt'})\n",
    "\n",
    "        # smoothed\n",
    "        yprime = gaussian_filter1d(\n",
    "            z['ave_mut-wt'],\n",
    "            sigma,\n",
    "            axis=-1,\n",
    "            order=0,\n",
    "            output=None,\n",
    "            mode='nearest',\n",
    "            cval=0.0,\n",
    "            truncate=4.0,\n",
    "        )\n",
    "\n",
    "        dt = pd.DataFrame({'time':z['time'] - 1, 'ave_mut-wt':yprime})\n",
    "        dt[gkey] = i\n",
    "\n",
    "        diffs = diffs.append(dt, ignore_index=True)\n",
    "        \n",
    "    # aesthetics\n",
    "    cmap_ctype={'GC': '#FAC18A',\n",
    "            'DCN': '#AAABAB',\n",
    "            'UBC': '#BA61BA',\n",
    "            'PC': '#EE5264',\n",
    "            'MLI1': '#F9EBAE',\n",
    "            'MLI2': '#88BB92',\n",
    "            'GoC': '#46A928',\n",
    "            'AS': '#F9AEAE',\n",
    "            'BG': '#AEB7F9',\n",
    "            'OPC': '#F1815F',\n",
    "            'OL': '#75A3B7',\n",
    "            'MG': '#AC5861',\n",
    "            'PER': '#2D284B',\n",
    "            'END': '#1C67EE'}\n",
    "\n",
    "\n",
    "    fig, ax = plt.subplots(1,1, figsize=(5,4))\n",
    "    sns.lineplot(x='time', y='ave_mut-wt', hue=gkey, size=None, \n",
    "                     style=None, data=diffs, palette=cmap_ctype, hue_order=None, \n",
    "                     hue_norm=None, sizes=None, size_order=None, size_norm=None, \n",
    "                     dashes=True, markers=None, style_order=None, \n",
    "                     units=None, estimator='mean', ci=95, n_boot=1000, seed=None, sort=True, \n",
    "                     err_style='band', err_kws=None, legend='brief', ax=ax, lw=3)\n",
    "    ax.legend(bbox_to_anchor=(1.1,1)).set_title('')\n",
    "    ax.set_xticks([0,19])\n",
    "    fig.savefig(os.path.join(pfp, 'dyn_diffs_ctypes.pdf'), bbox_inches='tight')\n",
    "    return diffs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## DCN\n",
    "\n",
    "dX_df = dyn_gene_diff(wtt, mutt, goi=cic_binding, gcname='DCN', normtype=None, pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/')\n",
    "#diffs = diffplot_dcn(dX_df, pfp='/home/cl2292/project/SCA1_snRNAseq/Mouse/results/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(pfp, exist_ok=True)\n",
    "goi=cic_binding\n",
    "results = pd.DataFrame()\n",
    "wt = wtt[(wtt.obs['sub4']=='DCN'), :].copy()\n",
    "mut = mutt[(mutt.obs['sub4']=='DCN'), :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "computing PCA\n",
      "    with n_comps=50\n",
      "    finished (0:00:13)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 50\n",
      "    finished (0:00:00)\n",
      "computing PCA\n",
      "    with n_comps=50\n",
      "    finished (0:00:03)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 50\n",
      "    finished (0:00:00)\n"
     ]
    }
   ],
   "source": [
    "X_wt, y_wt = meld_pseudotime(wt, goi=cic_binding)\n",
    "X_mut, y_mut = meld_pseudotime(mut, goi=cic_binding)\n",
    "y = np.concatenate((y_wt, y_mut))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# binning\n",
    "_, _, time_groupings = binned_statistic(y, y, bins=20)\n",
    "\n",
    "X = X_wt.append(X_mut)\n",
    "X = pd.DataFrame(zscore(X), index=X.index, columns=X.columns)\n",
    "X_wt = X.loc[X_wt.index, :]\n",
    "X_mut = X.loc[X_mut.index, :]\n",
    "\n",
    "X_wt['time_grouping'] = time_groupings[0:y_wt.shape[0]]\n",
    "X_wt = X_wt.groupby('time_grouping').mean().T\n",
    "X_mut['time_grouping'] = time_groupings[y_wt.shape[0]:]\n",
    "X_mut = X_mut.groupby('time_grouping').mean().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_mut[9] = (X_mut[10]+X_mut[8])/2\n",
    "X_wt[9] = (X_wt[10]+X_wt[8])/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "X_mut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "dXX = X_mut.subtract(X_wt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntimebins=20\n",
    "cm = plt.get_cmap('magma')\n",
    "timecolors = [cm(1.*i/ntimebins) for i in range(ntimebins)]\n",
    "# hmap1, WT\n",
    "p = sns.clustermap(\n",
    "    X_wt,\n",
    "    pivot_kws=None,\n",
    "    method='average',\n",
    "    metric='euclidean',\n",
    "    z_score=None,\n",
    "    standard_scale=None,\n",
    "    row_cluster=True,\n",
    "    col_cluster=False,\n",
    "    row_linkage=None,\n",
    "    figsize=(4, 20),\n",
    "    col_linkage=None,\n",
    "    row_colors=None,\n",
    "    col_colors=timecolors,\n",
    "    mask=None,\n",
    "    dendrogram_ratio=0.2,\n",
    "    colors_ratio=0.03,\n",
    "    cbar_pos=(1, 0, 0.1, 0.15),\n",
    "    tree_kws=None,\n",
    "    cmap='RdYlBu_r',\n",
    "    yticklabels=True,\n",
    "    xticklabels=False,\n",
    "    vmax=1, vmin=-1,\n",
    ")\n",
    "p.ax_heatmap.set_xlabel('')\n",
    "p.savefig(os.path.join(pfp, 'hmap_gexVtbin_wt_{}.pdf'.format('DCN')), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# hmap2, MUT\n",
    "p = sns.clustermap(\n",
    "    X_mut,\n",
    "    pivot_kws=None,\n",
    "    method='average',\n",
    "    metric='euclidean',\n",
    "    z_score=None,\n",
    "    standard_scale=None,\n",
    "    row_cluster=True,\n",
    "    col_cluster=False,\n",
    "    row_linkage=None,\n",
    "    figsize=(4, 20),\n",
    "    col_linkage=None,\n",
    "    row_colors=None,\n",
    "    col_colors=timecolors,\n",
    "    mask=None,\n",
    "    dendrogram_ratio=0.2,\n",
    "    colors_ratio=0.03,\n",
    "    cbar_pos=(1, 0, 0.1, 0.15),\n",
    "    tree_kws=None,\n",
    "    cmap='RdYlBu_r',\n",
    "    yticklabels=True,\n",
    "    xticklabels=False,\n",
    "    vmax=1, vmin=-1,\n",
    ")\n",
    "p.ax_heatmap.set_xlabel('')\n",
    "p.savefig(os.path.join(pfp, 'hmap_gexVtbin_mut_{}.pdf'.format('DCN')), bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# hmap3, DIFF\n",
    "dX  = X_mut.subtract(X_wt) # MUT - WT\n",
    "p_diff = sns.clustermap(\n",
    "        dX,\n",
    "        pivot_kws=None,\n",
    "        method='average',\n",
    "        metric='euclidean',\n",
    "        z_score=None,\n",
    "        standard_scale=None,\n",
    "        row_cluster=True,\n",
    "        col_cluster=False,\n",
    "        row_linkage=None,\n",
    "        figsize=(4, 20), \n",
    "        col_linkage=None,\n",
    "        row_colors=None,\n",
    "        col_colors=timecolors,\n",
    "        mask=None,\n",
    "        dendrogram_ratio=0.2,\n",
    "        colors_ratio=0.03,\n",
    "        cbar_pos=(1, 0, 0.1, 0.15),\n",
    "        tree_kws=None,\n",
    "        cmap='RdYlBu_r',\n",
    "        yticklabels=True,\n",
    "        xticklabels=False,\n",
    "        vmax=1, vmin=-1,\n",
    "    )\n",
    "p_diff.ax_heatmap.set_xlabel('')\n",
    "p_diff.savefig(os.path.join(pfp, 'hmap_gexVtbin_DIFF_{}.pdf'.format('DCN')), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# regplot\n",
    "gcname = 'DCN'\n",
    "x_diff_melted = dX.T.reset_index().melt(id_vars='time_grouping', var_name='GOI', value_name='mut-wt')\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,3))\n",
    "p = sns.regplot('time_grouping', 'mut-wt', data=x_diff_melted, \n",
    "            x_estimator=None, x_bins=None, x_ci='ci', scatter=True, \n",
    "            fit_reg=True, ci=95, n_boot=1000, units=None, seed=None, \n",
    "            order=1, logistic=False, lowess=True, robust=False, logx=False, x_partial=None, \n",
    "            y_partial=None, truncate=True, dropna=True, x_jitter=0.5, y_jitter=None, \n",
    "            label=None, color='gray', marker='o', scatter_kws={'s':1, 'alpha':0.6, 'lw':0}, line_kws=None, ax=ax)\n",
    "fig.savefig(os.path.join(pfp, 'regplot_mut-wtVtbin_{}.pdf'.format(gcname)), bbox_inches='tight')\n",
    "\n",
    "# histogram + smoothed line\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,3))\n",
    "z = dX.abs().mean(axis=0).reset_index()\n",
    "z = z.rename(columns={0:'ave_mut-wt'})\n",
    "sns.barplot(x='time_grouping', y='ave_mut-wt', data=z, color='gray', ax=ax)\n",
    "yprime = gaussian_filter1d(\n",
    "    z['ave_mut-wt'],\n",
    "    2,\n",
    "    axis=-1,\n",
    "    order=0,\n",
    "    output=None,\n",
    "    mode='nearest',\n",
    "    cval=0.0,\n",
    "    truncate=4.0,\n",
    ")\n",
    "ax.plot(z['time_grouping']-1, yprime, color='gray')\n",
    "fig.savefig(os.path.join(pfp, 'hist_mut-wtABSDIFF_{}'.format(gcname)), bbox_inches='tight')\n",
    "\n",
    "dX = dX.reset_index()\n",
    "dX['group'] = gcname\n",
    "dX_df = dX_df.append(dX, ignore_index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
