{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import os, sys, glob, re, math, pickle\n",
    "import phate, scprep, magic, meld\n",
    "import graphtools as gt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import time,random,datetime\n",
    "import networkx as nx\n",
    "import scvelo as scv\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from scipy import sparse\n",
    "from scipy.stats import mannwhitneyu, tiecorrect, rankdata\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import scanpy as sc\n",
    "from sklearn.dummy import DummyClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.cluster import SpectralClustering, OPTICS, cluster_optics_dbscan, AgglomerativeClustering\n",
    "from bbknn import bbknn\n",
    "import warnings\n",
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "import anndata2ri\n",
    "\n",
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# settings\n",
    "plt.rc('font', size = 8)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=8)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1\n",
    "\n",
    "# reproducibility\n",
    "rs = np.random.seed(42)\n",
    "\n",
    "# utils\n",
    "def mwu(X,Y,gene_names,correction=None,debug=False) :\n",
    "    '''\n",
    "    Benjamini-Hochberg correction implemented. Can change to Bonferonni\n",
    "    gene_names (list)\n",
    "    if X,Y single gene expression array, input x.reshape(-1,1), y.reshape(-1,1)\n",
    "    NOTE: get zeros sometimes because difference (p-value is so small)\n",
    "    '''\n",
    "    p=pd.DataFrame()\n",
    "    print('Mann-Whitney U w/Benjamini/Hochberg correction\\n')\n",
    "    start = time.time()\n",
    "    for i,g in enumerate(gene_names) :\n",
    "        if i==np.round(np.quantile(np.arange(len(gene_names)),0.25)) :\n",
    "            print('... 25% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.5)) :\n",
    "            print('... 50% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.75)) :\n",
    "            print('... 75% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        p.loc[i,'Gene']=g\n",
    "        if (tiecorrect(rankdata(np.concatenate((np.asarray(X[:,i]),np.asarray(Y[:,i])))))==0) :\n",
    "            if debug :\n",
    "                print('P-value not calculable for {}'.format(g))\n",
    "            p.loc[i,'pval']=np.nan\n",
    "        else :\n",
    "            _,p.loc[i,'pval']=mannwhitneyu(X[:,i],Y[:,i]) # continuity correction is True\n",
    "    print('... mwu computed in {:.2f}-s\\n'.format(time.time() - start))\n",
    "    # ignore NaNs, since can't do a comparison on these (change numbers for correction)\n",
    "    p_corrected = p.loc[p['pval'].notna(),:]\n",
    "    new_pvals = multipletests(p_corrected['pval'],method='fdr_bh')\n",
    "    p_corrected['pval_corrected'] = new_pvals[1]\n",
    "    return p_corrected\n",
    "\n",
    "def log2aveFC(X,Y,gene_names,AnnData=None) :\n",
    "    '''not sensitivity to directionality due to subtraction\n",
    "    X and Y full arrays, subsetting performed here\n",
    "    `gene_names` (list): reduced list of genes to calc\n",
    "    `adata` (sc.AnnData): to calculate reduced list. NOTE: assumes X,Y drawn from adata.var_names\n",
    "    '''\n",
    "    if not AnnData is None :\n",
    "        g_idx = [i for i,g in enumerate(AnnData.var_names) if g in gene_names]\n",
    "        fc=pd.DataFrame({'Gene':AnnData.var_names[g_idx],\n",
    "                         'log2FC':np.log2(X[:,g_idx].mean(axis=0)) - np.log2(Y[:,g_idx].mean(axis=0))}) # returns NaN if negative value \n",
    "    else :\n",
    "        fc=pd.DataFrame({'Gene':gene_names,\n",
    "                         'log2FC':np.log2(X.mean(axis=0)) - np.log2(Y.mean(axis=0))})\n",
    "    return fc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fps\n",
    "dfp = '/home/cl2292/project/SCA1_snRNAseq/Human/data/'\n",
    "pfp = '/home/cl2292/project/SCA1_snRNAseq/Human/results_20230402/'\n",
    "pdfp = '/home/cl2292/project/SCA1_snRNAseq/Human/data/processed/'\n",
    "sc.settings.figdir = pfp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak memory: 8163.16 MiB, increment: 932.71 MiB\n",
      "loaded @230503.10:16:40\n",
      "took 6.69-s to load data\n",
      "peak memory: 9450.94 MiB, increment: 1287.77 MiB\n",
      "loaded @230503.10:16:48\n",
      "took 8.24-s to load data\n"
     ]
    }
   ],
   "source": [
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='230502_Ctrl_sampling_UBC_PC_MLI1_MLI2_GoC.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit wt = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))\n",
    "    \n",
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='230502_SCA1_sampling_UBC_PC_MLI1_MLI2_GoC.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit mut = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MLI1    1331\n",
      "MLI2    1088\n",
      "PC       370\n",
      "GoC       44\n",
      "UBC       43\n",
      "Name: sub4, dtype: int64\n",
      "MLI2    1960\n",
      "MLI1    1684\n",
      "PC       224\n",
      "GoC      105\n",
      "UBC       78\n",
      "Name: sub4, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(wt.obs.sub4.value_counts())\n",
    "print(mut.obs.sub4.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating all\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "UBC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:43\n",
      "    Ncells in Y:78\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.67-s\n",
      "... 50% completed in 43.80-s\n",
      "... 75% completed in 73.23-s\n",
      "... mwu computed in 109.19-s\n",
      "\n",
      "... computed in 118.25-s\n",
      "\n",
      "PC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:370\n",
      "    Ncells in Y:224\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 20.38-s\n",
      "... 50% completed in 45.36-s\n",
      "... 75% completed in 75.87-s\n",
      "... mwu computed in 112.75-s\n",
      "\n",
      "... computed in 119.25-s\n",
      "\n",
      "MLI1, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:1331\n",
      "    Ncells in Y:1684\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 25.97-s\n",
      "... 50% completed in 56.41-s\n",
      "... 75% completed in 92.31-s\n",
      "... mwu computed in 134.62-s\n",
      "\n",
      "... computed in 144.77-s\n",
      "\n",
      "MLI2, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:1088\n",
      "    Ncells in Y:1960\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 25.77-s\n",
      "... 50% completed in 56.06-s\n",
      "... 75% completed in 91.82-s\n",
      "... mwu computed in 134.15-s\n",
      "\n",
      "... computed in 144.35-s\n",
      "\n",
      "GoC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:44\n",
      "    Ncells in Y:105\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.51-s\n",
      "... 50% completed in 43.60-s\n",
      "... 75% completed in 73.21-s\n",
      "... mwu computed in 109.18-s\n",
      "\n",
      "... computed in 113.23-s\n",
      "\n",
      "Finished timepoint all in 10.66-min\n",
      "DGE finished in 10.69-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB4, IMP, Sampling, 'UBC','PC','MLI1','MLI2','GoC'\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub4'\n",
    "    fname = 'hum_imp_samp_UBC_PC_MLI1_MLI2_GoC' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['all'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in ['UBC','PC','MLI1','MLI2','GoC'] :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SCA1'.format(i))\n",
    "            print('----')\n",
    "            X = wt[((wt.obs[group]==i)), :].layers['imputed']\n",
    "            Y = mut[((mut.obs[group]==i)), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            #dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_WTup.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_WTdown.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating all\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "UBC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:43\n",
      "    Ncells in Y:78\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 18.97-s\n",
      "... 50% completed in 42.45-s\n",
      "... 75% completed in 71.59-s\n",
      "... mwu computed in 107.10-s\n",
      "\n",
      "... computed in 114.23-s\n",
      "\n",
      "PC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:370\n",
      "    Ncells in Y:224\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.09-s\n",
      "... 50% completed in 42.83-s\n",
      "... 75% completed in 72.19-s\n",
      "... mwu computed in 108.07-s\n",
      "\n",
      "... computed in 112.62-s\n",
      "\n",
      "MLI1, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:1331\n",
      "    Ncells in Y:1684\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 22.16-s\n",
      "... 50% completed in 48.94-s\n",
      "... 75% completed in 81.21-s\n",
      "... mwu computed in 119.98-s\n",
      "\n",
      "... computed in 127.53-s\n",
      "\n",
      "MLI2, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:1088\n",
      "    Ncells in Y:1960\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 21.91-s\n",
      "... 50% completed in 48.35-s\n",
      "... 75% completed in 80.23-s\n",
      "... mwu computed in 118.69-s\n",
      "\n",
      "... computed in 126.26-s\n",
      "\n",
      "GoC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:44\n",
      "    Ncells in Y:105\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 18.93-s\n",
      "... 50% completed in 42.37-s\n",
      "... 75% completed in 71.16-s\n",
      "... mwu computed in 106.54-s\n",
      "\n",
      "... computed in 110.52-s\n",
      "\n",
      "Finished timepoint all in 9.85-min\n",
      "DGE finished in 9.87-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB4, Non_IMP, Sampling, 'UBC','PC','MLI1','MLI2','GoC'\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub4'\n",
    "    fname = 'hum_no_imp_samp_UBC_PC_MLI1_MLI2_GoC' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['all'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in ['UBC','PC','MLI1','MLI2','GoC'] :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SCA1'.format(i))\n",
    "            print('----')\n",
    "            X = wt.X[(wt.obs[group]==i), :]\n",
    "            Y = mut.X[(mut.obs[group]==i), :]\n",
    "            \n",
    "            if sparse.issparse(X):\n",
    "                X = X.todense()\n",
    "            if sparse.issparse(Y):\n",
    "                Y = Y.todense()\n",
    "                \n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "            \n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            #dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_WTup.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_WTdown.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak memory: 2939.37 MiB, increment: 2635.01 MiB\n",
      "loaded @230503.10:49:02\n",
      "took 3.49-s to load data\n",
      "peak memory: 7085.73 MiB, increment: 4146.36 MiB\n",
      "loaded @230503.10:49:07\n",
      "took 4.70-s to load data\n"
     ]
    }
   ],
   "source": [
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='230502_Ctrl_sampling_AS BG OPC OL MG PER END.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit wt = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))\n",
    "    \n",
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='230502_SCA1_sampling_AS BG OPC OL MG PER END.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit mut = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BG     2445\n",
      "OL     2367\n",
      "AS     1648\n",
      "OPC     713\n",
      "MG      638\n",
      "END     365\n",
      "PER      73\n",
      "Name: sub4, dtype: int64\n",
      "BG     4384\n",
      "AS     3367\n",
      "OL     1946\n",
      "MG     1315\n",
      "END     967\n",
      "OPC     879\n",
      "PER     254\n",
      "Name: sub4, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(wt.obs.sub4.value_counts())\n",
    "print(mut.obs.sub4.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating all\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "AS, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:1648\n",
      "    Ncells in Y:3367\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 30.69-s\n",
      "... 50% completed in 65.95-s\n",
      "... 75% completed in 106.82-s\n",
      "... mwu computed in 153.97-s\n",
      "\n",
      "... computed in 168.50-s\n",
      "\n",
      "BG, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:2445\n",
      "    Ncells in Y:4384\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 34.81-s\n",
      "... 50% completed in 73.92-s\n",
      "... 75% completed in 118.71-s\n",
      "... mwu computed in 169.78-s\n",
      "\n",
      "... computed in 188.82-s\n",
      "\n",
      "OPC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:713\n",
      "    Ncells in Y:879\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 22.76-s\n",
      "... 50% completed in 50.14-s\n",
      "... 75% completed in 83.14-s\n",
      "... mwu computed in 122.60-s\n",
      "\n",
      "... computed in 129.52-s\n",
      "\n",
      "OL, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:2367\n",
      "    Ncells in Y:1946\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 28.43-s\n",
      "... 50% completed in 61.42-s\n",
      "... 75% completed in 100.03-s\n",
      "... mwu computed in 145.11-s\n",
      "\n",
      "... computed in 157.88-s\n",
      "\n",
      "MG, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:638\n",
      "    Ncells in Y:1315\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 23.37-s\n",
      "... 50% completed in 51.34-s\n",
      "... 75% completed in 85.12-s\n",
      "... mwu computed in 125.32-s\n",
      "\n",
      "... computed in 132.95-s\n",
      "\n",
      "PER, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:73\n",
      "    Ncells in Y:254\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.96-s\n",
      "... 50% completed in 44.53-s\n",
      "... 75% completed in 74.89-s\n",
      "... mwu computed in 111.49-s\n",
      "\n",
      "... computed in 116.03-s\n",
      "\n",
      "END, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:365\n",
      "    Ncells in Y:967\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 22.05-s\n",
      "... 50% completed in 48.72-s\n",
      "... 75% completed in 81.06-s\n",
      "... mwu computed in 119.86-s\n",
      "\n",
      "... computed in 126.21-s\n",
      "\n",
      "Finished timepoint all in 17.00-min\n",
      "DGE finished in 17.03-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB4, IMP, Sampling, 'AS','BG','OPC','OL','MG','PER','END'\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub4'\n",
    "    fname = 'hum_imp_samp_AS_BG_OPC_OL_MG_PER_END' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['all'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in ['AS','BG','OPC','OL','MG','PER','END'] :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SCA1'.format(i))\n",
    "            print('----')\n",
    "            X = wt[((wt.obs[group]==i)), :].layers['imputed']\n",
    "            Y = mut[((mut.obs[group]==i)), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            #dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_WTup.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_WTdown.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating all\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "AS, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:1648\n",
      "    Ncells in Y:3367\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 23.75-s\n",
      "... 50% completed in 51.99-s\n",
      "... 75% completed in 85.62-s\n",
      "... mwu computed in 125.44-s\n",
      "\n",
      "... computed in 137.72-s\n",
      "\n",
      "BG, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:2445\n",
      "    Ncells in Y:4384\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 25.43-s\n",
      "... 50% completed in 55.53-s\n",
      "... 75% completed in 91.01-s\n",
      "... mwu computed in 132.40-s\n",
      "\n",
      "... computed in 144.75-s\n",
      "\n",
      "OPC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:713\n",
      "    Ncells in Y:879\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.99-s\n",
      "... 50% completed in 44.52-s\n",
      "... 75% completed in 74.61-s\n",
      "... mwu computed in 110.92-s\n",
      "\n",
      "... computed in 116.50-s\n",
      "\n",
      "OL, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:2367\n",
      "    Ncells in Y:1946\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 22.63-s\n",
      "... 50% completed in 49.83-s\n",
      "... 75% completed in 82.64-s\n",
      "... mwu computed in 121.59-s\n",
      "\n",
      "... computed in 130.49-s\n",
      "\n",
      "MG, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:638\n",
      "    Ncells in Y:1315\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 20.30-s\n",
      "... 50% completed in 45.19-s\n",
      "... 75% completed in 75.46-s\n",
      "... mwu computed in 112.23-s\n",
      "\n",
      "... computed in 118.22-s\n",
      "\n",
      "PER, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:73\n",
      "    Ncells in Y:254\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 18.55-s\n",
      "... 50% completed in 41.75-s\n",
      "... 75% completed in 70.43-s\n",
      "... mwu computed in 105.46-s\n",
      "\n",
      "... computed in 109.77-s\n",
      "\n",
      "END, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:365\n",
      "    Ncells in Y:967\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.61-s\n",
      "... 50% completed in 43.75-s\n",
      "... 75% completed in 73.43-s\n",
      "... mwu computed in 109.26-s\n",
      "\n",
      "... computed in 114.61-s\n",
      "\n",
      "Finished timepoint all in 14.53-min\n",
      "DGE finished in 14.56-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB4, Non_IMP, Sampling, 'AS','BG','OPC','OL','MG','PER','END'\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub4'\n",
    "    fname = 'hum_no_imp_samp_AS_BG_OPC_OL_MG_PER_END' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['all'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in ['AS','BG','OPC','OL','MG','PER','END'] :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SCA1'.format(i))\n",
    "            print('----')\n",
    "            X = wt.X[(wt.obs[group]==i), :]\n",
    "            Y = mut.X[(mut.obs[group]==i), :]\n",
    "            \n",
    "            if sparse.issparse(X):\n",
    "                X = X.todense()\n",
    "            if sparse.issparse(Y):\n",
    "                Y = Y.todense()\n",
    "                \n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "            \n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            #dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_WTup.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_WTdown.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak memory: 15633.19 MiB, increment: 15331.92 MiB\n",
      "loaded @230503.12:28:23\n",
      "took 15.11-s to load data\n",
      "peak memory: 31084.44 MiB, increment: 15451.24 MiB\n",
      "loaded @230503.12:28:39\n",
      "took 15.53-s to load data\n"
     ]
    }
   ],
   "source": [
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='230503_Ctrl_sampling_GC50000.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit wt = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))\n",
    "    \n",
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='230503_SCA1_sampling_GC50000.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit mut = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC    50000\n",
      "Name: sub4, dtype: int64\n",
      "GC    50000\n",
      "Name: sub4, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(wt.obs.sub4.value_counts())\n",
    "print(mut.obs.sub4.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating all\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "GC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:50000\n",
      "    Ncells in Y:50000\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 372.54-s\n",
      "... 50% completed in 750.65-s\n",
      "... 75% completed in 1141.34-s\n",
      "... mwu computed in 1553.64-s\n",
      "\n",
      "... computed in 1947.20-s\n",
      "\n",
      "Finished timepoint all in 32.45-min\n",
      "DGE finished in 32.46-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB4, IMP, Sampling, 'GC'\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub4'\n",
    "    fname = 'hum_imp_samp_GC' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['all'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in ['GC'] :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SCA1'.format(i))\n",
    "            print('----')\n",
    "            X = wt[((wt.obs[group]==i)), :].layers['imputed']\n",
    "            Y = mut[((mut.obs[group]==i)), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            #dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_WTup.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_WTdown.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating all\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "GC, WT vs SCA1\n",
      "----\n",
      "    Ncells in X:50000\n",
      "    Ncells in Y:50000\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 122.30-s\n",
      "... 50% completed in 249.43-s\n",
      "... 75% completed in 381.25-s\n",
      "... mwu computed in 517.81-s\n",
      "\n",
      "... computed in 674.33-s\n",
      "\n",
      "Finished timepoint all in 11.24-min\n",
      "DGE finished in 11.24-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB4, Non_IMP, Sampling, 'GC'\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub4'\n",
    "    fname = 'hum_no_imp_samp_GC' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['all'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in ['GC'] :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SCA1'.format(i))\n",
    "            print('----')\n",
    "            X = wt.X[(wt.obs[group]==i), :]\n",
    "            Y = mut.X[(mut.obs[group]==i), :]\n",
    "            \n",
    "            if sparse.issparse(X):\n",
    "                X = X.todense()\n",
    "            if sparse.issparse(Y):\n",
    "                Y = Y.todense()\n",
    "                \n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "            \n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=wt.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(Y,X,wt.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            #dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_WTup.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_WTdown.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
